sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala

```diff
@@ -1382,11 +1407,59 @@ case class DropIndex(
     copy(table = newChild)
 }
 
-case class TableSpec(
+trait TableSpec {
+  def properties: Map[String, String]
+  def provider: Option[String]
+  def location: Option[String]
+  def comment: Option[String]
+  def serde: Option[SerdeInfo]
+  def external: Boolean
+  def withNewLocation(newLocation: Option[String]): TableSpec
+  def treePatternBits: BitSet
+}
+
+case class UnresolvedTableSpec(
+    properties: Map[String, String],
+    provider: Option[String],
+    optionsExpressions: Map[String, Expression],
```

Comment:
@dtenedor I got a new idea about this.
How about we have a new expression for `optionsExpressions`
```
case class UnresolvedTableOptions(options: Seq[(String, Expression)])
  extends Expression with Unevaluable {
  override def nullable: Boolean = true

  override def dataType: DataType = MapType[StringType, StringType]

  override def children: Seq[Expression] = optionsExpressions.map(_._2)

  override protected def withNewChildrenInternal(
    newChildren: IndexedSeq[Expression]): Expression = {
      assert(options.length == newChildren.length)
      ...
      }
  }
}
```
So that we don't need to handle the treePatternBits or hacking the method `mapExpressions`. 
Note that here the Options is `Seq` instead of `Map` to make sure the expression order is fixed.

Reply:
Thanks for the good suggestion, this makes it simpler! This is done.